src/rule.jl

```diff
@@ -221,7 +237,8 @@ julia> r(2 * (a+b+c))
 
 **Predicates**:
 
-Predicates can be used on both `~x` and `~~x` by using the `~x::f` or `~~x::f`.
+There are two kinds of predicates, namely over slot variables and over the whole rule.
+For the former, predicates can be used on both `~x` and `~~x` by using the `~x::f` or `~~x::f`.
```

Comment:
I'm thinking we should maybe deprecate `~x::f`...

It's kind of a confusing syntax, compare:

```julia
@rule ~x::(x->x isa Integer) => 0
@rule ~x => 0 where ~x isa Integer
```

@YingboMa ?

Reply:
Oh never mind actually it is more efficient to use `::` because it stops parsing preemptively.